renovation_core_graphql/tags/resolvers.py

```diff
@@ -0,0 +1,38 @@
+from graphql import GraphQLResolveInfo
+
+import frappe
+from frappe.desk.doctype.tag.tag import add_tag, remove_tag, get_tags
+
+
+def add_tag_link_resolver(obj, info: GraphQLResolveInfo, **kwargs):
+    input = kwargs.get("input")
+    doctype = input.get("doctype")
+    docname = input.get("docname")
+    tag = input.get("tag")
+    return add_tag(tag=tag, dt=doctype, dn=docname)
+
+
+def remove_tag_link_resolver(obj, info: GraphQLResolveInfo, **kwargs):
+    input = kwargs.get("input")
+    doctype = input.get("doctype")
+    docname = input.get("docname")
+    tag = input.get("tag")
+    remove_tag(tag=tag, dt=doctype, dn=docname)
+    return True
+
+
+def get_tags_resolver(obj, info: GraphQLResolveInfo, **kwargs):
+    doctype = kwargs.get("doctype")
+    search = kwargs.get("search") or ""
+
+    # This will return all tags in the system
+    if not doctype:
+        return get_tags(doctype, search)
+
+    # This will return all tags for a specific doctype
+    query = """
+        SELECT DISTINCT tag FROM `tabTag Link`
+         WHERE parenttype = %s AND tag LIKE %s
+        """
+    tags = frappe.db.sql(query, [doctype, "%" + search + "%"], as_dict=True, debug=1)
```

Comment:
if not doctype:
        return get_tags(doctype, search)
that part check permissions since get_list is used..
so
i suggest to do `frappe.has_permission(doctype,"read")` ?

also can the query part be replaced with `frappe.desk.doctype.tag.tag.get_tagged_docs`

Reply:
For the part `frappe.desk.doctype.tag.tag.get_tagged_docs`, that will return the documents that have a particular tag, however, what we want here is the tags in general used against this doctype (or all if doctype isn't specified).

As for permission, there's actually no need. I think I'll replace `get_tags` with our own query because if you notice, the `doctype` argument is unused in `get_tags` 😅